File: trunk/JAVA/PokerProtocol/src/protocol/GameTCPServer.java
package protocol;

import game.Card;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

import poker.game.MoneyPot;
import poker.game.PlayerInfo;
import poker.game.PokerGame;
import poker.game.PokerGame.TypeState;
import poker.game.TableInfo;
import poker.game.TypeAction;
import poker.game.TypeRound;
import poker.game.observer.PokerGameAdapter;
import protocol.commands.DisconnectCommand;
import protocol.commands.ICommand;
import protocol.commands.game.BetTurnEndedCommand;
import protocol.commands.game.BetTurnStartedCommand;
import protocol.commands.game.GameEndedCommand;
import protocol.commands.game.GameStartedCommand;
import protocol.commands.game.PlayerHoleCardsChangedCommand;
import protocol.commands.game.PlayerJoinedCommand;
import protocol.commands.game.PlayerLeftCommand;
import protocol.commands.game.PlayerMoneyChangedCommand;
import protocol.commands.game.PlayerPlayMoneyCommand;
import protocol.commands.game.PlayerTurnBeganCommand;
import protocol.commands.game.PlayerTurnEndedCommand;
import protocol.commands.game.PlayerWonPotCommand;
import protocol.commands.game.TableClosedCommand;
import protocol.commands.game.TableInfoCommand;
import protocol.observer.game.GameServerAdapter;
import protocol.observer.game.GameServerObserver;

public class GameTCPServer implements Runnable
{
    // Communication Things
    private boolean m_isConnected = true;
    private final BlockingQueue<String> m_incoming = new LinkedBlockingQueue<String>();
    private final BlockingQueue<String> m_outcoming = new LinkedBlockingQueue<String>();
    
    // POKER Things
    private final PlayerInfo m_player;
    private final PokerGame m_game;
    private final GameServerObserver m_commandObserver = new GameServerObserver();
    
    /**
     * Create a new player
     * 
     * @param p_name
     *            The name of the player
     * @param p_money
     *            The starting chips of the player
     * @param p_socket
     *            The TCP socket with the client
     * @throws IOException
     *             Error with the TCP connection
     */
    public GameTCPServer(PokerGame game, String name, int money)
    {
        m_game = game;
        m_player = new PlayerInfo(name, money);
        initializeCommandObserver();
    }
    
    public String getOutcoming() throws InterruptedException
    {
        return m_outcoming.take();
    }
    
    public void incoming(String s) throws InterruptedException
    {
        m_incoming.put(s);
    }
    
    protected String receive() throws InterruptedException
    {
        final String line = m_incoming.take();
        m_commandObserver.receiveSomething(line);
        return line;
    }
    
    public boolean canStartGame()
    {
        return m_player.canPlay();
    }
    
    /**
     * Send a message to the client
     * 
     * @param p_msg
     *            The message to send
     */
    protected void send(ICommand command)
    {
        // System.out.println("Server SEND TO " + this.m_name + " [" + p_msg + "]");
        try
        {
            m_outcoming.put(command.encodeCommand());
        }
        catch (final InterruptedException e)
        {
            e.printStackTrace();
        }
    }
    
    public PlayerInfo getPlayer()
    {
        return m_player;
    }
    
    public PokerGame getGame()
    {
        return m_game;
    }
    
    public boolean joinGame()
    {
        initializePokerObserver();
        return m_game.joinGame(m_player);
    }
    
    public void sitIn()
    {
        send(new TableInfoCommand(m_game.getTable(), m_player));
        m_game.sitInGame(m_player);
    }
    
    public boolean isConnected()
    {
        return m_isConnected;
    }
    
    @Override
    public void run()
    {
        try
        {
            while (isConnected())
            {
                receive();
            }
        }
        catch (final InterruptedException e)
        {
            return;
        }
    }
    
    public void start()
    {
        new Thread(this).start();
    }
    
    private void initializePokerObserver()
    {
        m_game.attach(new PokerGameAdapter()
        {
            @Override
            public void gameBettingRoundEnded(TypeRound r)
            {
                final List<MoneyPot> pots = m_game.getTable().getPots();
                final ArrayList<Integer> amounts = new ArrayList<Integer>();
                for (final MoneyPot pot : pots)
                {
                    amounts.add(pot.getAmount());
                }
                
                for (int i = pots.size(); i < m_game.getTable().getNbMaxSeats(); i++)
                {
                    amounts.add(0);
                }
                send(new BetTurnEndedCommand(amounts, r));
            }
            
            @Override
            public void playerHoleCardsChanged(PlayerInfo p)
            {
                final Card[] holeCards = p.getCards(p.getNoSeat() == m_player.getNoSeat());
                send(new PlayerHoleCardsChangedCommand(p.getNoSeat(), holeCards[0].getId(), holeCards[1].getId(), p.isPlaying()));
            }
            
            @Override
            public void gameEnded()
            {
                send(new GameEndedCommand());
            }
            
            @Override
            public void playerWonPot(PlayerInfo p, MoneyPot pot, int wonAmount)
            {
                send(new PlayerWonPotCommand(p.getNoSeat(), pot.getId(), wonAmount, p.getMoneySafeAmnt()));
            }
            
            @Override
            public void playerActionTaken(PlayerInfo p, TypeAction reason, int playedAmount)
            {
                send(new PlayerTurnEndedCommand(p.getNoSeat(), p.getMoneyBetAmnt(), p.getMoneySafeAmnt(), m_game.getTable().getTotalPotAmnt(), reason, playedAmount, p.isPlaying()));
            }
            
            @Override
            public void playerMoneyChanged(PlayerInfo p)
            {
                send(new PlayerMoneyChangedCommand(p.getNoSeat(), p.getMoneySafeAmnt()));
            }
            
            @Override
            public void everythingEnded()
            {
                send(new TableClosedCommand());
                m_isConnected = false;
            }
            
            @Override
            public void playerActionNeeded(PlayerInfo p, PlayerInfo last)
            {
                send(new PlayerTurnBeganCommand(p.getNoSeat(), last.getNoSeat()));
            }
            
            @Override
            public void gameBlindsNeeded()
            {
                send(new GameStartedCommand(m_game.getTable().getNoSeatDealer(), m_game.getTable().getNoSeatSmallBlind(), m_game.getTable().getNoSeatBigBlind()));
            }
            
            @Override
            public void gameBettingRoundStarted()
            {
                final Card[] cards = new Card[5];
                m_game.getTable().getCards().toArray(cards);
                for (int i = 0; i < 5; ++i)
                {
                    if (cards[i] == null)
                    {
                        cards[i] = Card.NO_CARD;
                    }
                }
                send(new BetTurnStartedCommand(cards[0].getId(), cards[1].getId(), cards[2].getId(), cards[3].getId(), cards[4].getId(), m_game.getRound()));
            }
            
            @Override
            public void playerJoined(PlayerInfo p)
            {
                send(new PlayerJoinedCommand(p.getNoSeat(), p.getName(), p.getMoneySafeAmnt()));
            }
            
            @Override
            public void playerLeaved(PlayerInfo p)
            {
                send(new PlayerLeftCommand(p.getNoSeat()));
            }
        });
    }
    
    private void initializeCommandObserver()
    {
        m_commandObserver.subscribe(new GameServerAdapter()
        {
            @Override
            public void disconnectCommandReceived(DisconnectCommand command)
            {
                m_isConnected = false;
                m_player.setZombie();
                final TableInfo t = m_game.getTable();
                if (m_game.getState() == TypeState.PLAYERS_WAITING)
                {
                    m_game.leaveGame(m_player);
                }
                else if (t.getNoSeatCurrPlayer() == m_player.getNoSeat())
                {
                    if (t.canCheck(m_player))
                    {
                        m_game.playMoney(m_player, 0);
                    }
                    else
                    {
                        m_game.playMoney(m_player, -1);
                    }
                }
            }
            
            @Override
            public void commandReceived(String command)
            {
                System.out.println("<TcpPokerClient :" + m_player.getName() + "> RECV [" + command + "]");
            }
            
            @Override
            public void playMoneyCommandReceived(PlayerPlayMoneyCommand command)
            {
                m_game.playMoney(m_player, command.getPlayed());
            }
        });
    }
}

File: C#/Com.Ericmas001.Game.Poker.Protocol.Server/GameTCPServer.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.IO;
using System.Threading;
using Com.Ericmas001.Game.Poker.Protocol.Commands;
using Com.Ericmas001.Game.Poker.Protocol.Commands.Game;
using Com.Ericmas001.Games;
using Com.Ericmas001.Net.Protocol.JSON;
using Com.Ericmas001.Util;
using Com.Ericmas001.Game.Poker.DataTypes.Enums;
using Com.Ericmas001.Game.Poker.DataTypes.EventHandling;
using Com.Ericmas001.Game.Poker.DataTypes;
using Com.Ericmas001.Game.Poker.Logic;
using Com.Ericmas001.Net.Protocol;
using Com.Ericmas001.Game.Poker.DataTypes.Parameters;

namespace Com.Ericmas001.Game.Poker.Protocol.Server
{
    public class GameTCPServer : CommandQueueCommunicator<GameObserver>
    {
        #region Fields
        private readonly PlayerInfo m_Player;
        private readonly PokerGame m_Game;
        private readonly int m_ID;
        private readonly UserInfo m_UserInfo;
        #endregion Fields

        #region Events
        public event EventHandler<KeyEventArgs<int>> LeftTable = delegate { };
        #endregion Events

        #region Properties
        public int ID { get { return m_ID; } }
        public PlayerInfo Player { get { return m_Player; } }
        public PokerGame Game { get { return m_Game; } }
        #endregion Properties

        #region Ctors & Init
        public GameTCPServer(int id, PokerGame game, string name, int money)
        {
            m_ID = id;
            m_Game = game;
            m_Player = new PlayerInfo(name, 0);
            m_UserInfo = null;
            base.SendedSomething += OnServerSendedSomething;

        }

        public GameTCPServer(int id, PokerGame game, UserInfo userInfo)
        {
            m_ID = id;
            m_Game = game;
            m_UserInfo = userInfo;
            m_Player = new PlayerInfo(m_UserInfo.DisplayName, 0);
        }

        private void InitializePokerObserver()
        {
            m_Game.Observer.GameBettingRoundEnded += OnGameBettingRoundEnded;
            m_Game.Observer.PlayerHoleCardsChanged += OnPlayerHoleCardsChanged;
            m_Game.Observer.GameEnded += OnGameEnded;
            m_Game.Observer.PlayerWonPot += OnPlayerWonPot;
            m_Game.Observer.PlayerActionTaken += OnPlayerActionTaken;
            m_Game.Observer.PlayerMoneyChanged += OnPlayerMoneyChanged;
            m_Game.Observer.EverythingEnded += OnEverythingEnded;
            m_Game.Observer.PlayerActionNeeded += OnPlayerActionNeeded;
            m_Game.Observer.GameBlindNeeded += OnGameBlindNeeded;
            m_Game.Observer.GameBettingRoundStarted += OnGameBettingRoundStarted;
            m_Game.Observer.GameGenerallyUpdated += OnGameGenerallyUpdated;
            m_Game.Observer.PlayerJoined += OnPlayerJoined;
            m_Game.Observer.SeatUpdated += OnSeatUpdated;
            m_Game.Observer.PlayerLeft += OnPlayerLeft;
        }

        protected override void InitializeCommandObserver()
        {
            m_CommandObserver.CommandReceived += OnCommandReceived;
            m_CommandObserver.DisconnectCommandReceived += OnDisconnectCommandReceived;
            m_CommandObserver.PlayMoneyCommandReceived += OnPlayMoneyCommandReceived;
            m_CommandObserver.SitInCommandReceived += OnSitInCommandReceived;
            m_CommandObserver.SitOutCommandReceived += OnSitOutCommandReceived;
        }
        #endregion Ctors & Init

        #region GameServer Event Handling
        void OnServerSendedSomething(object sender, KeyEventArgs<string> e)
        {
            LogManager.Log(LogLevel.MessageLow, "OnServerSendedSomething", "<Game:{0}> SEND [{1}]", m_Player.Name, e.Key);
        }
        #endregion GameServer Event Handling

        #region PokerObserver Event Handling
        void OnGameBettingRoundEnded(object sender, RoundEventArgs e)
        {
            List<MoneyPot> pots = m_Game.Table.Pots;
            List<int> amounts = pots.Select(p => p.Amount).ToList();

            for (int i = pots.Count; i < m_Game.Table.Params.MaxPlayers; i++)
                amounts.Add(0);

            Send(new BetTurnEndedCommand() 
            { 
                PotsAmounts = amounts, 
                Round = e.Round,
            });
        }

        void OnPlayerHoleCardsChanged(object sender, PlayerInfoEventArgs e)
        {
            PlayerInfo p = e.Player;
            GameCard[] holeCards = p.NoSeat == m_Player.NoSeat ? p.Cards : p.RelativeCards;

            Send(new PlayerHoleCardsChangedCommand() 
            {
                PlayerPos = p.NoSeat, 
                State = p.State,
                CardsID = holeCards.Select(c => c.Id).ToList() ,
            });
        }

        void OnGameEnded(object sender, EventArgs e)
        {
            Send(new GameEndedCommand());
        }

        void OnPlayerWonPot(object sender, PotWonEventArgs e)
        {
            PlayerInfo p = e.Player;
            Send(new PlayerWonPotCommand()
            {
                PlayerPos = p.NoSeat,
                PotID = e.Id,
                Shared = e.AmountWon,
                PlayerMoney = p.MoneySafeAmnt,
            });
        }

        void OnPlayerActionTaken(object sender, PlayerActionEventArgs e)
        {
            PlayerInfo p = e.Player;
            Send(new PlayerTurnEndedCommand()
            {
                PlayerPos = p.NoSeat,
                PlayerBet = p.MoneyBetAmnt,
                PlayerMoney = p.MoneySafeAmnt,
                TotalPot = m_Game.Table.TotalPotAmnt,
                ActionType = e.Action,
                ActionAmount = e.AmountPlayed,
                State = p.State,
            });
        }

        void OnPlayerMoneyChanged(object sender, PlayerInfoEventArgs e)
        {
            PlayerInfo p = e.Player;
            Send(new PlayerMoneyChangedCommand()
            {
                PlayerPos = p.NoSeat,
                PlayerMoney = p.MoneySafeAmnt,
            });
        }

        void OnEverythingEnded(object sender, EventArgs e)
        {
            Send(new TableClosedCommand());
            m_IsConnected = false;
        }

        void OnPlayerActionNeeded(object sender, HistoricPlayerInfoEventArgs e)
        {
            Send(new PlayerTurnBeganCommand()
            {
                PlayerPos = e.Player.NoSeat,
                LastPlayerNoSeat = e.Last.NoSeat,
                MinimumRaise = m_Game.Table.MinimumRaiseAmount,
            });
        }

        void OnGameBlindNeeded(object sender, EventArgs e)
        {
            Send(new GameStartedCommand() { NeededBlind = m_Game.GameTable.GetBlindNeeded(m_Player) });
        }

        void OnGameBettingRoundStarted(object sender, RoundEventArgs e)
        {
            Send(new BetTurnStartedCommand()
            {
                Round = e.Round,
                CardsID = m_Game.Table.Cards.Select(x => x.Id).ToList()
            });
        }

        private void OnGameGenerallyUpdated(object sender, EventArgs e)
        {
            SendTableInfo();
        }

        void OnPlayerJoined(object sender, PlayerInfoEventArgs e)
        {
            PlayerInfo p = e.Player;
            Send(new PlayerJoinedCommand()
            {
                PlayerName = p.Name,
            });
        }

        void OnSeatUpdated(object sender, SeatEventArgs e)
        {
            if (!e.Seat.IsEmpty && m_Player.NoSeat != e.Seat.NoSeat)
                e.Seat.Player.HoleCards = e.Seat.Player.RelativeCards;

            Send(new SeatUpdatedCommand()
            {
                Seat = e.Seat,
            });
        }

        void OnPlayerLeft(object sender, PlayerInfoEventArgs e)
        {
            PlayerInfo p = e.Player;
            Send(new PlayerLeftCommand()
            {
                PlayerPos = p.NoSeat,
            });
        }
        #endregion PokerObserver Event Handling

        #region CommandObserver Event Handling
        void OnPlayMoneyCommandReceived(object sender, CommandEventArgs<PlayerPlayMoneyCommand> e)
        {
            m_Game.PlayMoney(m_Player, e.Command.Played);
        }

        void OnSitOutCommandReceived(object sender, CommandEventArgs<PlayerSitOutCommand> e)
        {
            Send(e.Command.EncodeResponse(m_Game.SitOut(m_Player)));
        }

        void OnSitInCommandReceived(object sender, CommandEventArgs<PlayerSitInCommand> e)
        {
            TableParams parms = m_Game.Table.Params;
            int money;
            if(parms.Lobby.OptionType == LobbyTypeEnum.Training)
                money = ((LobbyOptionsTraining)parms.Lobby).StartingAmount;
            else
            {
                money = e.Command.MoneyAmount;
                if (m_UserInfo == null || m_UserInfo.TotalMoney < money || money < parms.LimitedMinimumBuyIn || (parms.LimitMaximumBuyIn && money > parms.LimitedMaximumBuyIn))
                {
                    Send(e.Command.EncodeResponse(-1));
                    return;
                }
                m_UserInfo.TotalMoney -= money;
            }

            m_Player.MoneySafeAmnt = money;
            SeatInfo seat = m_Game.GameTable.AskToSitIn(m_Player, e.Command.NoSeat);
            Send(e.Command.EncodeResponse(seat == null ? -1 : seat.NoSeat));
            if( seat != null)
                m_Game.SitIn(m_Player);
        }

        void OnDisconnectCommandReceived(object sender, CommandEventArgs<DisconnectCommand> e)
        {
            if (m_UserInfo != null && m_Game.Params.Lobby.OptionType == LobbyTypeEnum.Career)
                m_UserInfo.TotalMoney += m_Player.MoneySafeAmnt;

            LeftTable(this, new KeyEventArgs<int>(m_ID));

            m_IsConnected = false;
            m_Player.State = PlayerStateEnum.Zombie;

            TableInfo t = m_Game.Table;
            LogManager.Log(LogLevel.Message, "GameServer.m_CommandObserver_DisconnectCommandReceived", "> Client '{0}' left table: {2}:{1}", m_Player.Name, t.Params.TableName, m_ID);

            if (m_Game.State == GameStateEnum.WaitForPlayers)
                m_Game.LeaveGame(m_Player);
            else if (t.NoSeatCurrentPlayer == m_Player.NoSeat)
            {
                if (t.CanCheck(m_Player))
                    m_Game.PlayMoney(m_Player, 0);
                else
                    m_Game.PlayMoney(m_Player, -1);
            }
        }

        void OnCommandReceived(object sender, StringEventArgs e)
        {
            LogManager.Log(LogLevel.MessageLow, "OnCommandReceived", "<Game:{0}> RECV [{1}]", m_Player.Name, e.Str);
        }
        #endregion CommandObserver Event Handling

        #region Public Methods
        public bool CanStartGame
        {
            get
            {
                return m_IsConnected && m_Player.CanPlay;
            }
        }
        public void SendTableInfo()
        {
            TableInfoCommand cmd = new TableInfoCommand();
            cmd.GameHasStarted = m_Game.IsPlaying;
            PokerTable table = m_Game.GameTable;
            lock (table)
            {
                PlayerInfo playerSendingTo = m_Player;

                cmd.BoardCardIDs = table.Cards.Select(c => c.Id).ToList();
                cmd.Seats = new List<SeatInfo>();

                cmd.Params = table.Params;

                cmd.TotalPotAmount = table.TotalPotAmnt;
                cmd.NbPlayers = cmd.Params.MaxPlayers;

                cmd.PotsAmount = table.Pots.Select(pot => pot.Amount).ToList();
                cmd.PotsAmount.AddRange(Enumerable.Repeat(0, cmd.Params.MaxPlayers - table.Pots.Count));

                for (int i = 0; i < cmd.Params.MaxPlayers; ++i)
                {
                    SeatInfo si = new SeatInfo() { NoSeat = i };
                    cmd.Seats.Add(si);
                    SeatInfo gameSeat = table.Seats[i];
                    if (gameSeat.IsEmpty)
                        continue;
                    si.Player = gameSeat.Player.Clone();

                    //If we are not sending the info about the player who is receiving, don't show the cards unless you can
                    if (i != playerSendingTo.NoSeat && si.Player.IsPlaying)
                        si.Player.HoleCards = gameSeat.Player.RelativeCards;

                    if (si.Player.HoleCards.Length != 2)
                        si.Player.HoleCards = new GameCard[] { GameCard.NO_CARD, GameCard.NO_CARD };

                    si.Attributes = gameSeat.Attributes;
                }
            }
            Send(cmd);
        }

        public bool JoinGame()
        {
            InitializePokerObserver();
            return m_Game.JoinGame(m_Player);
        }
        #endregion Public Methods
    }
}
